### ISSUE TITLE
Stabilize Time to Read block

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#53776]
The Time to Read block was implemented in #43403 and shipped into the Gutenberg plugin in February 2023, but is not yet bundled with the core.

Issues like #13796, #15350 have been reported related to this block, but [all issues labeled with this block are closed and there are no open issues](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue+label%3A%22%5BBlock%5D+Time+to+Read%22+). Therefore, I'm hoping this block will be included in WordPress 6.4.

However, this block relies on the new `wp_word_count()` function to calculate the number of characters on the server side. A patch has been submitted in [the core ticket](https://core.trac.wordpress.org/ticket/57987) for this new function, but in order to ship this block in the core, I believe this function needs to be implemented in the core first.

**Comment by annezazu** [comment]
Wanted to flag this feedback from the [6.4 roadmap](https://make.wordpress.org/core/2023/08/22/roadmap-to-6-4/#comment-45485) from @joedolson:

>The “time to read” is something I’m not very comfortable including in core. I have some fundamental reservations about this because the idea that there’s a certain amount of time a post should take to read is inherently ableist. It can create a sense of inadequacy in people who struggle with reading by placing a “normal” value to an experience that real people experience in radically different ways.

@scruffian brought up the idea of a Word Count option instead which might be a better path to pursue for Core.

**Comment by t-hamano** [comment]
>The “time to read” is something I’m not very comfortable including in core. I have some fundamental reservations about this because the idea that there’s a certain amount of time a post should take to read is inherently ableist. It can create a sense of inadequacy in people who struggle with reading by placing a “normal” value to an experience that real people experience in radically different ways.

I can understand this point. However, there is also a Time to Read section in the Document Panel of the post editor. This section was added in #41611. This feature is included in Gutenberg 13.7, so this item has existed since WordPress 6.1.

![time-to-read](https://github.com/WordPress/gutenberg/assets/54422211/dc7c98fa-c94f-46e1-89a6-990617882eb9)

**Comment by torounit** [comment]
The same is true for the "Word Count" already included in Gutenberg, but the "Word Count Block" does not work for languages that do not separate words with spaces, such as CJKV.

**Comment by scruffian** [comment]
What about character count? How does time to read cope with CJKV languages?

**Comment by t-hamano** [comment]
> How does time to read cope with CJKV languages?

This block calculates the number of characters or words, taking into account the site language count type (`words` | `characters_excluding_spaces` | `characters_including_spaces`) first, in order to calculate the reading time. It simply [divides that number by the "average rate of reading per minute"](https://github.com/WordPress/gutenberg/blob/255ac93af14c658122383964cddfabd0f87ae00e/packages/block-library/src/post-time-to-read/edit.js#L66-L71).

If this block were to be changed to one that only outputs the number of characters or words in the content, I think this could be accomplished by simply removing just the division logic.

Alternatively, it might be a good idea to provide the following UI so that the site creator can change to a count type other than that site language, as shown below.

![word count type](https://github.com/WordPress/gutenberg/assets/54422211/8a9776e8-04c6-430c-8418-6b318a28a225)

Also, if it is problematic for this block to enforce "normal" standards, we can provide a UI with variable reading rates as follows.

![reading-rate](https://github.com/WordPress/gutenberg/assets/54422211/c90357d1-3e46-44c3-9290-d8998bb9e8e4)

**Comment by joedolson** [comment]
> I can understand this point. However, there is also a Time to Read section in the Document Panel of the post editor. This section was added in https://github.com/WordPress/gutenberg/pull/41611. This feature is included in Gutenberg 13.7, so this item has existed since WordPress 6.1.

Having this data available as a point of reference for content authors and editors is profoundly different from having it present for site visitors. I'm not particularly thrilled with it in the editor, either, but in that context it's only directed at people who are actively working on the document, and can be used as a relative metric to assess the length of your working drafts; the value is more abstract in that context, because the audience is not, in fact, reading the document.

**Comment by annezazu** [comment]
I think we need more time to determine adding this block to Core and let discussion continue further. Going to punt to 6.5 at this point to focus current efforts.

**Comment by t-hamano** [comment]
Thanks for the detailed explanation, @joedolson!

I would love to hear from @afercia and @alexstine , who have a deep background in accessibility, about their opinions on this discussion 🙏

**Comment by alexstine** [comment]
Time to read is always very subjective. It can be based on how much you know about the subject, learning disabilities, ability to consume content, etc. For example, I am sure the time to read wouldn't take into account how fast I can read something at 4X with my screen reader. Not sure word count would even communicate a time to read type feature. I see it as not really an experience improvement.

**Comment by afercia** [comment]
I can only second what Joe and Alex already pointed out.

The reading time is highly subjective by its own nature. It also varies by age and education level. Such a tool can only vaguely indicate the _average_ reading time, which may not be useful for all users.. Also, I wonder how the algorithm works with languages where a single character represents a whole word or concept.

That said, after some research, it seems there's some data that show length indication is beneficial. Most notably it seems that:
- engagement rates improve
- bounce rate drops down
- time spent on the site increases

WordPress is in the position to set new best practices, when it's beneficial. I wonder if some research and testing could help identify a better way to indicate an article length that is not based on average time.

**Comment by t-hamano** [comment]
Thank you all for your input. Based on your input, we might be able to make a decision not to ship this block to the core.

If so, will this block remain on the Gutenberg plugin? As far as I know, all blocks that have been developed with the Gutenberg plugin are meant to be shipped to the core, and I'm not aware of any blocks that have been decided not to ship 🤔

**Comment by masteradhoc** [comment]
@t-hamano i would love to get this block into core. you can see this information on a lot websites and it definately lets the user know in a first place what to expect inside this article. This seems a UX-Improvement for me. If your a fast / slow reader you probably know - and adapt. 

I would see it the same as a hiking track - you dont go without an indication. But if your a fast/slow hiker you'll know that it will take you less / more.

I second also @afercia's benefits. Please keep working on this :)

**Comment by MaggieCabrera** [comment]
I arrived here because I tried to add this block to TT4 but I saw it's still not in core. Hopefully, we can add it to the next default theme then!

**Comment by afercia** [comment]
> I wonder how the algorithm works with languages where a single character represents a whole word or concept.

> WordPress is in the position to set new best practices, when it's beneficial. I wonder if some research and testing could help identify a better way to indicate an article length that is not based on average time.

I'd like to point out both the above points are very important.

I'm not sure the average reading time would be appropriate for all languages, as it is mostly based on Western languages.

I wonder how that would work for languages where single characters are either whole concepts (ideographic symbols) or phonetic symbols (syllables). I'm afraid a reading time based on the number of characters would miserably fail in these cases.

I'm not an expert of these languages (although I would love to learn at least some basics) so I'd defer to someone else with better linguistics knowledge than me.

**Comment by peteringersoll** [comment]
I really don't like the idea of specific times. People read and comprehend at different rates. To imply "If you don't read this in two minutes you are slow" or even worse is unnecessary. Has it been discussed to compromise with description like short, medium, long, very long, etc.? Then we have the indicator without injecting an ability.

**Comment by t-hamano** [comment]
[The WordPress 6.5 beta 1 is scheduled for release on February 13th](https://make.wordpress.org/core/2024/01/19/editor-important-milestones-for-wordpress-6-5/), so we may want to consider making a decision on this issue.

Personally, I would be happy to see it shipped to Core, as I was involved in the development of this block. However, the consensus among accessibility experts is that this block should not be included in the core. I am of course satisfied with the opinion, so I think it is okay to close this issue.

If you have a user who really needs a block like this, you can enable the Gutenberg plugin or create a custom block.

**Comment by masteradhoc** [comment]
I'd love to get this further and merged to the next wp version. I see this as a block that people can use in case they want. So people that prefer a solution like "short,long,.." or prefer no times at all dont need to use it.

Minutes will never be exact as everyone is a bit different. But still its a nice hint on how long the article will be.
This leads to a better UX as the user knows what he can expect when he clicks the post.

Just to recap again:
"I would see it the same as a hiking track - you dont go without an indication. But if your a fast/slow hiker you'll know that it will take you less / more."
see https://github.com/WordPress/gutenberg/issues/53776#issuecomment-1715068572

Also @afercia's points are quite crucial and i saw this same on some of our client sides. Would be great if this work can be merged.
"That said, after some research, it seems there's some data that show length indication is beneficial. Most notably it seems that:
- engagement rates improve
- bounce rate drops down
- time spent on the site increases"

Thanks @t-hamano

**Comment by jamiemarsland** [comment]
I would really love to see this in the next wp version - the hiking analogy works for me, and it would be something that would really highlight why folks should try block themes.  The https://www.nasa.gov/ is a great example of where it works well.

**Comment by peteringersoll** [comment]
I've unsuccessfully tried to find actual studies about the improvements. There are articles that state there are studies, but I could not find the studies themselves. Does anyone know where I could find those? 

Using the hiking analogy, I think it still follows that there are better ways to indicate the length of an article. Hiking trails around here indicate a length (miles, km), and difficulty (easy/paved, moderate, challenging) - but not time.

**Comment by jamiemarsland** [comment]
Good points Peter - but for me the number 1 reason to use the time to read block is to encourage people to click and read. Time does this for me whereas number of words, or something vague wouldn't.

**Comment by joedolson** [comment]
I think that the difficulty with including a number of words is that most people don't know how fast they read, so knowing that an article has 900 words doesn't really tell them anything. A time to read may be horribly inaccurate, but it's easier to understand what it means. 

I'm not sure there is any way of providing a measurement that's accurate when we don't know anything about the potential reader, so any value we use has to be considered an approximation. 

I like that this block rounds aggressively, ignoring seconds - I wonder if it would be best to switch to `math.ceil()` so it always rounds up?

Here's an interesting [study about reading speed in different languages](https://irisreading.com/average-reading-speed-in-various-languages/). This is possibly information we should be taking into consideration - there are some significant differences in these numbers.

**Comment by peteringersoll** [comment]
I'll bow out. Each site owner can decide whether to use or not use. What I really want to see is real A/B testing to have actual data to demonstrate the value, effectiveness, etc. This topic, and other interface-related topics, often refer to studies - without actually citing those studies so others can understand the data gathering methods. But that's a different discussion. PS. I'd rather see effort going into improving the table block.

**Comment by colorful-tones** [comment]
I feel like this is plugin territory. I would vote to close this out and not merge.

That is difficult to say, because I know that @t-hamano put effort into this. 🤗

**Comment by t-hamano** [comment]
Thank you everyone for your opinions.

I think this issue can be removed from the editor task board, or milestone. This issue has been punted once from WP6.4 to 6.5. Deeper discussion and research will be needed to find the ideal approach, so even if we punt it to WP6.6 this time, it will be difficult to make a decision on whether to include it in the core or not, and we might need to punt it further.

I think it's okay to leave this issue open and include it in a milestone when we find a ideal approach to ship to the core. In any case, users who really want to use this block just need to enable the Gutenberg plugin.

**Comment by jamiemarsland** [comment]
I created a little video about this to get some wider community feedback - the comments are interesting
https://youtu.be/apvDdOb_Vtg?si=R1fZvRLOkSwFMeGp

**Comment by masteradhoc** [comment]
Thanks @jamiemarsland for the great video! 

Just saw the comments and unbelievable how good the feedback is towards the time to read block. Even many people commented that indicate it helps them as indication although they know it will take them slightly longer.

Hope we can push the great work into core soon.

**Comment by annezazu** [comment]
Thanks for flagging this for the comments! I do want to note that it's not fair to say this has been axed when this is actively under discussion, especially with this new proposal https://github.com/WordPress/gutenberg/issues/58773 Just in case anyone else is reading this and think it's been fully "axed". There's more to the story, especially when thinking about core maintained blocks folks can install but that aren't included by default with Core.

**Comment by jamiemarsland** [comment]
Thanks for clarifying Anne - fyi the context of the video was that it's been axed from the core WordPress install, which it has - hopefully the comments will help in the discussion if the Time to read block is considered for a core supported non core block. A quick follow on question, do you know when the time to read block is scheduled to be removed from the Gutenberg plugin?  In my testing today the Time to Read Block is broken in Version 17.9.0 of Gutenberg.

**Comment by annezazu** [comment]
Update for you all that there's an effort underway to have the [time to read block as a canonical block](https://github.com/WordPress/gutenberg/pull/61504). T[his is part of a larger effort to have Core created and supported blocks](https://github.com/WordPress/gutenberg/issues/58773) in the block directory that folks can install but don't come bundled into Core. Rather than stabilizing the time to read block, it looks like this is the more viable direction.

**Comment by colorful-tones** [comment]
Hi folks,
We are only one week away from the Beta 1 cut-off date for [WordPress 6.6](https://make.wordpress.org/core/6-6/). This issue hasn’t seen any movement in a while, so we (the editor triage leads of the 6.6 release) have decided to remove it from the [WordPress 6.6 Editor Tasks project board](https://github.com/orgs/WordPress/projects/179/views/1).

**Comment by pbearne** [comment]
refreshed the wordcount function patch 
https://core.trac.wordpress.org/ticket/57987#comment:7

**Comment by pbearne** [comment]
moved time to read block from experimental

**Comment by t-hamano** [comment]
@pbearne, I think there is no consensus whether we should ship this block to core. Let's close this PR and continue the discussion.

As you can see from the beginning of this issue, the most important thing is accessibility. Unless this is resolved, I don't think we can ship this block into core.

The following two issues are also very important:

- https://github.com/WordPress/gutenberg/issues/58773
- https://github.com/WordPress/gutenberg/issues/71026

---

One thing I regret is that I should have developed this block as `core/post-content-count` instead of `core/time-to-read`, which might have made it more flexible. For example:

- Users can choose what to count.
- Users can choose the word count type if they want to show the content amount.
- Users can override the average reading rate if they want to show the time to read the content.

The image of how it works is as follows:

https://github.com/user-attachments/assets/abd26ed9-85ec-4fd9-9ab2-901964e11a8f

**Comment by pbearne** [comment]
@t-hamano, is there any code for the post-content-count version?
If not, if I create this, is it likely to be added?

**Comment by t-hamano** [comment]
> is there any code for the post-content-count version?

No, that's just a temporary example I created.

> If not, if I create this, is it likely to be added?

Creating a block doesn't mean it will ship to core; it's something that needs to be discussed and decided upon.

Another important thing to note is that the Time to Read block cannot be removed from Gutenberg.

**Comment by scruffian** [comment]
> it's something that needs to be discussed and decided upon.
IMO, [this issue](https://github.com/WordPress/gutenberg/issues/71026) makes it clear that the block is wanted in core, so we should stabilize this in core.

I think it would also be good to explore other options for how to improve this block:
https://github.com/WordPress/gutenberg/issues/71583

**Comment by t-hamano** [comment]
> The “time to read” is something I’m not very comfortable including in core. I have some fundamental reservations about this because the idea that there’s a certain amount of time a post should take to read is inherently ableist. It can create a sense of inadequacy in people who struggle with reading by placing a “normal” value to an experience that real people experience in radically different ways.

@joedolson if we could add more options than just "Time to Read" I think we could ship this block to Core. What do you think?

**Comment by peteringersoll** [comment]
Agree about being not assuming reading ability. It should not be that difficult to come up with a version that, at least, allows options for indicating length in more inclusive ways.

**Comment by jeryj** [comment]
tl;dr: I'm leaning toward the most flexible and inclusive option being:

- Expose Word Reading speed as a number value so sites can change this value
- Add a toggle so there can be a display as range option
- Add descriptions to explain the accessibility considerations/impact of these fields

Rationale:

[Each language has its own average word reading speed](https://irisreading.com/average-reading-speed-in-various-languages/), so our fixed value won't be accurate, nor will the range.

Also, this "average" is not an average of the population, it's an **average of normally-sighted people aged 18 - 35**:

> Methods.: Ten texts were designed for 17 languages each by a linguist who matched content, length, difficulty, and linguistic complexity. The texts then were used to assess reading speeds of 436 normally-sighted native speakers (age 18–35 years, 25 per language, 36 in Japanese), presented at a distance of 40 cm and size 1 M, that is 10-point Times New Roman font. Reading time (aloud) was measured by stopwatch.
> https://iovs.arvojournals.org/article.aspx?articleid=2166061&_ga=2.57419498.1109406412.1757588380-247723523.1757588380

They also mention an average for readers aged 60–85 years, which is about 20% slower than the 18-35 group:

> Age also has a role for reading speed. In this study, age 18 to 35 years was an inclusion criterion. We showed in a previous study using IReST 6 (first edition) that older subjects (aged 60–85 years) read more slowly by a factor of approximately 20% (English and German readers). 
> https://iovs.arvojournals.org/article.aspx?articleid=2166061&_ga=2.57419498.1109406412.1757588380-247723523.1757588380

When looking at readers with age-related macular degeneration, the average of English readers was 38 words/min:

> The reading speeds in English and German native-speaking patients with age-related macular degeneration (AMD) were assessed in two studies using the IReST (first edition), when patients read the texts with appropriate magnification. A total of 40 English AMD patients had a mean reading speed of 0.44 ± 0.39 log char/sec, that is 38 words/min, viewed at their critical print size. 31 In 530 German AMD patients, the mean reading speed was 20 ± 33 words/min before and 72 ± 35 words/min after adaptation of optimal magnifying visual aids. 32  
> https://iovs.arvojournals.org/article.aspx?articleid=2166061&_ga=2.57419498.1109406412.1757588380-247723523.1757588380

What do we do with all of this information?  The "Time to Read" block is a reasonable feature that WordPress users will likely want. They see it on their favorite sites. We should offer this feature to keep WordPress relevant to the general site user. However, I also get uneasy about the idea of endorsing this block in core, when the concept is inherently ableist.

To balance these, I think we can offer an educational opportunity in the block, similar to Alt text for images. Hopefully that will give people who want the Time to Read block an opportunity to understand its implications and make the choice that's best for them, as well as help them determine what value to use for their site's average reading speed. For example, an ophthalmologist specializing in treating macular degeneration patients will have an audience with a different reading rate. 

**I'm by no means set on this and am new to this block and concepts, but based on the research, I'd like to see:**

- "Average reading speed" as a customizable input. 
- Display as range as a toggle. The value used should lean towards the slower value, so not Average +/- 20%, but Average + 30%. I think this helps reduce the impact to the marginalized readers.
- Some kind of educational description explaining the accessibility considerations when configuring/choosing to use this block.

**Comment by peteringersoll** [comment]
Also: Cognitive abilities, dyslexia, other vision issues etc. all impact a person's reading rate. Thanks for providing real-world info on this.

**Comment by t-hamano** [comment]
My one suggestion would be to offer the option to display "word count" or "character count." This is a fact that applies to everyone, unlike "time to read."

However, since this is the "Time" to Read block, I'm not sure we can introduce options other than "time" 😅

**Comment by jeryj** [comment]
@t-hamano I agree that the word count would be the most inclusive if it was common enough for us to know what it means in practice 😅 

It's too bad there isn't the concept of "pages" for length on the web. If an article is 1 page vs 20, I'd have a good idea of how much time to set aside. If it says 500 vs 10,000 words, I'd have less of an idea.

**Comment by joedolson** [comment]
Pages aren't that useful; they depend hugely on too many other factors in publishing physical items, anyway. But publishers use 250-300 words per page as a baseline estimate for authoring guidance.

But I agree that trying to provide an extrapolation in pages is something that wouldn't be seriously meaningful for most users.

Word count could be meaningful, just as a general range. I may not be able to literally understand the difference between 1000 words and 5000 words, but I can get a sense of scale.

I would argue that this can still have non-time based options within the "Time to Read" concept. The block is about providing users with a mechanism to figure out how long something might take them. A word count is also a way of providing that. It definitely shouldn't be the default, but could be available as an option.

Thanks for your detailed comments, @jeryj - I think what you've proposed makes a lot of sense. I think we'll always be better off estimating on the slower end. Readers are more likely to be frustrated if they thought they would have time for an article, and didn't, than if they finish the article sooner than expected.

**Comment by pbearne** [comment]
Hi all

The time to read a block is a requested feature!
I suspect that the reading speed is always going to be wrong, but it's better than not having an indication of how long it will take

I feel that should be a site-wide value, and would like to suggest adding a filter around $average_reading_rate

**Comment by t-hamano** [comment]
Based on the discussion so far, I would like to restate my proposal to implement the following features in order to ship this block in core:

- The reading time option is the default
  - [x] Add a slider to adjust the average speed: #71632
  - [x] Add a toggle to show a range of reading times, rather than one time: #71606
- [x] Add an option to display word or character count instead of reading time https://github.com/WordPress/gutenberg/issues/71583
   - [ ] Add an option to change the word count type: Words, Characters (excluding spaces), Characters (including spaces)

**Comment by pbearne** [comment]
> * Add a slider to adjust the average speed:

Let's make sure the default is in a filter or maybe a global setting as well.

I feel I would like to set this for all content on site, not have to adjust per post(content)

**Thinking about this**
Can we have the slider be +/- from the global/auto setting
So if we adjust the global value so the reading time track from that (increase for challenging reading and reduce if easy content)

**Comment by joedolson** [comment]
Agree with @pbearne; this adjustment really needs to be something that can be adjusted globally, or it will be very labor intensive for users. The occasional per-article adjustment could happen, but in most cases this will probably be site-wide.

**Comment by masteradhoc** [comment]
Fully agree here as well that this is mostly a global adjustment thats needed. For me personally a filter would be enough - but if we can (as an enhancement, maybe v2) a global setting would be fine as well.

**Comment by t-hamano** [comment]
> Let's make sure the default is in a filter or maybe a global setting as well.

As far as I know, there is no single global setting for configuring specific attributes of a specific block, and I would be reluctant to add a new global setting just for this block.

The default average reading speed is planned to be defined as a part of block.json, so it should be easy to change the default value by using `register_block_type_args` or `block_type_metadata_settings` filter.

**Comment by pbearne** [comment]
If needed we can create small plugin to allow easy comtrol of the reading speed

**Comment by t-hamano** [comment]
It's looking more likely that we'll ship this block to core, so let's continue enhancing it and evaluate whether we can ship it in 6.9.

**Comment by ntsekouras** [comment]
Would a new setting (maybe an editor setting? 🤔 ) make sense to also be used from other places we calculate reading time, like the `outline`?

 So, what's is missing to stabilize this block? From what I've read so far one suggestion is:

>Add an option to display word or character count instead of reading time

But I'd argue that for now, having a hardcoded list to choose from or even a text input that overrides the dynamic calculation of reading time, might be enough? I hope I'm not missing lot's of context..

**Comment by scruffian** [comment]
> But I'd argue that for now, having a hardcoded list to choose from or even a text input that overrides the dynamic calculation of reading time, might be enough? I hope I'm not missing lot's of context..

This is already possible by filtering the block attribute.

> Add an option to display word or character count instead of reading time

There is a PR for this:

https://github.com/WordPress/gutenberg/pull/71841#issuecomment-3323686394

**Comment by t-hamano** [comment]
Let's take another look at the latest specifications of this block.

<img width="802" height="397" alt="Image" src="https://github.com/user-attachments/assets/74263dc6-bdac-4d34-ae81-50c22164e272" />

- Time is displayed as a range by default. The default value for average reading speed is defined in block.json, and this value can be changed with a PHP filter.
- There is a toggle option to show time or word count, but these toggles may be changed to a single-select option. See https://github.com/WordPress/gutenberg/pull/71897
